pkg/startup/autodetection/ipv4/poolselector.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package ipv4
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	FALLBACK_IPPOOL_TEMPLATE = "172.%d.0.0/16"
+	FALLBACK_IPPOOL_MIN      = 16
+	FALLBACK_IPPOOL_MAX      = 31
+)
+
+type poolSelector struct {
+	actualPool    *net.IPNet
+	poolCounter   int
+	hostAddresses []net.IPNet
+}
+
+var hostIPAddressRetriever func() ([]net.IPNet, error) = retrieveIPAddresses
+
+// GetDefaultIPv4Pool detects host interfaces and selects default IP pool without overlapping
+func GetDefaultIPv4Pool(preferedPool string) string {
+	log.Debug("Auto-detecting IPv4 interfaces to select default pool")
+
+	_, preferedNet, err := net.ParseCIDR(preferedPool)
+	if err != nil {
+		log.Errorf("Unable to parse prefered pool: '%s'", err.Error())
```

Comment:
I think we should be returning errors from this function and having the calling code handle them. Presumably if, somehow, we fail to parse this CIDR, we don't want the calling code to continue on using the bad CIDR, right?